File: mobile/lib/models/book_loan.dart
class BookLoan {
  final int id;
  final String loanReference;
  final String studentId;
  final String? studentName;
  final String bookIsbn;
  final String? bookTitle;
  final String loanDate; // yyyy-MM-dd
  final String dueDate; // yyyy-MM-dd
  final String? returnDate; // yyyy-MM-dd (null if not returned)
  final String status; // 'BORROWED', 'RETURNED', 'OVERDUE', 'LOST'
  final double fineAmount;

  BookLoan({
    required this.id,
    required this.loanReference,
    required this.studentId,
    this.studentName,
    required this.bookIsbn,
    this.bookTitle,
    required this.loanDate,
    required this.dueDate,
    this.returnDate,
    required this.status,
    required this.fineAmount,
  });

  bool get isReturned => status == 'RETURNED';
  bool get isOverdue => status == 'OVERDUE';

  factory BookLoan.fromJson(Map<String, dynamic> json) {
    final retDateVal = json['returnDate'] as String?;
    final retDate =
        (retDateVal == null ||
            retDateVal.trim().isEmpty ||
            retDateVal == 'null')
        ? null
        : retDateVal;

    return BookLoan(
      id: json['id'] as int,
      loanReference: json['loanReference'] as String,
      studentId: json['studentId'] as String,
      studentName: json['studentName'] as String?,
      bookIsbn: json['bookIsbn'] as String,
      bookTitle: json['bookTitle'] as String?,
      loanDate: json['loanDate'] as String,
      dueDate: json['dueDate'] as String,
      returnDate: retDate,
      status: json['status'] as String,
      fineAmount: (json['fineAmount'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'loanReference': loanReference,
      'studentId': studentId,
      'studentName': studentName,
      'bookIsbn': bookIsbn,
      'bookTitle': bookTitle,
      'loanDate': loanDate,
      'dueDate': dueDate,
      'returnDate': returnDate,
      'status': status,
      'fineAmount': fineAmount,
    };
  }
}

File: desktop/model/BookLoan.java
package model;

public class BookLoan {
    private int id;
    private String loanReference;
    private String studentId;
    private String studentName;
    private String bookIsbn;
    private String bookTitle;
    private String loanDate;
    private String dueDate;
    private String returnDate;
    private String status; // BORROWED, RETURNED, OVERDUE, LOST
    private double fineAmount;

    public BookLoan() {}

    public BookLoan(int id, String loanReference, String studentId, String studentName, String bookIsbn, String bookTitle, String loanDate, String dueDate, String returnDate, String status, double fineAmount) {
        this.id = id;
        this.loanReference = loanReference;
        this.studentId = studentId;
        this.studentName = studentName;
        this.bookIsbn = bookIsbn;
        this.bookTitle = bookTitle;
        this.loanDate = loanDate;
        this.dueDate = dueDate;
        this.returnDate = returnDate;
        this.status = status;
        this.fineAmount = fineAmount;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getLoanReference() {
        return loanReference;
    }

    public void setLoanReference(String loanReference) {
        this.loanReference = loanReference;
    }

    public String getStudentId() {
        return studentId;
    }

    public void setStudentId(String studentId) {
        this.studentId = studentId;
    }

    public String getStudentName() {
        return studentName;
    }

    public void setStudentName(String studentName) {
        this.studentName = studentName;
    }

    public String getBookIsbn() {
        return bookIsbn;
    }

    public void setBookIsbn(String bookIsbn) {
        this.bookIsbn = bookIsbn;
    }

    public String getBookTitle() {
        return bookTitle;
    }

    public void setBookTitle(String bookTitle) {
        this.bookTitle = bookTitle;
    }

    public String getLoanDate() {
        return loanDate;
    }

    public void setLoanDate(String loanDate) {
        this.loanDate = loanDate;
    }

    public String getDueDate() {
        return dueDate;
    }

    public void setDueDate(String dueDate) {
        this.dueDate = dueDate;
    }

    public String getReturnDate() {
        return returnDate;
    }

    public void setReturnDate(String returnDate) {
        this.returnDate = returnDate;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public double getFineAmount() {
        return fineAmount;
    }

    public void setFineAmount(double fineAmount) {
        this.fineAmount = fineAmount;
    }
}
